pkg/migration-tool/reader/path_reader.go

```diff
@@ -0,0 +1,169 @@
+package reader
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/prometheus/prometheus/pkg/labels"
+	"github.com/prometheus/prometheus/prompb"
+	"github.com/prometheus/prometheus/tsdb"
+	"github.com/prometheus/prometheus/tsdb/chunkenc"
+	tsdb_errors "github.com/prometheus/prometheus/tsdb/errors"
+	"github.com/timescale/promscale/pkg/log"
+	plan "github.com/timescale/promscale/pkg/migration-tool/planner"
+)
+
+type PathRead struct {
+	c            context.Context
+	path         string
+	plan         *plan.Plan
+	sigBlockRead chan *plan.Block // To the writer.
+	sigForceStop chan struct{}
+}
+
+// NewPathRead creates a reader that reads from local Prometheus blocks.
+func NewPathRead(c context.Context, path string, p *plan.Plan, sigRead chan *plan.Block, needsStopChan bool) (Reader, error) {
+	reader := &PathRead{
+		c:            c,
+		path:         path,
+		plan:         p,
+		sigBlockRead: sigRead,
+	}
+	if needsStopChan {
+		reader.sigForceStop = make(chan struct{})
+	}
+	return reader, nil
+}
+
+func (rr *PathRead) SigStop() {
+	rr.sigForceStop <- struct{}{}
+}
+
+// Run runs the remote read and starts fetching the samples from the read storage.
+func (rr *PathRead) Run(errChan chan<- error) {
```

Comment:
I think this function need a refactoring/cleanup. Its very hard to follow. I suggest moving some functions define in here out to the toplevel as a start.

Reply:
Could you re-look this comment now as the recent commits have made some funcs top-level, so if this holds true even now.

Reply:
Looks much better now